[flang][debug] Improve handling of dummy character arguments.

abidh opened the pull request:
As described in #107998, we were not handling the case well when length of the character is not part of the type. This PR handles one of the case when the length can be calculated by looking at the result of corresponding `fir.unboxchar`.

The DIStringTypeAttr have a `stringLength` field that can be a variable. We create an artificial variable that will hold the length and used as value of `stringLength` field.

As we don't have any `fircg.ext_declare` to which we can attach the location. This PR attaches it to `fir.unboxchar` and we check for it during codegen and generate a `DbgValueOp`. I thought about the idea of generating a `fircg.ext_declare` so that we can attach the variable to it. That will avoid touching `fir.unboxchar` in codegen. But it would also have required adding an `alloca` to hold the value so that we can use `DbgDeclareOp`. I opted for `fir.unboxchar` as it seemed simpler.

Fixes #107998.

abidh committed:
[flang][debug] Improve handling of dummy character arguments.

As described in #107998, we were not handling the case well when
length of the character is not part of the type. This PR handles one of
the case when the length can be calculated by looking at the result of
corresponding `fir.unboxchar`.

The DIStringTypeAttr have a `stringLength` field that can be a variable.
We create an artificial variable that will hold the length and used
as value of `stringLength` field.

As we don't have any `fircg.ext_declare` to which we can attach the
location. This PR attaches it to fir.unboxchar and we check for it
during CodeGen and generate a `DbgValueOp`. I thought about the idea of
generating a fircg.ext_declare so that we can attach the variable to it.
That will avoid touching `fir.unboxchar` in codegen. But it would also
have required adding an `alloca` to hold the value so that we can use
`DbgDeclareOp`. I opted for `fir.unboxchar`` as it seemed simpler.

jinisusan commented on flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:
```diff
@@ -286,7 +287,25 @@ mlir::LLVM::DITypeAttr DebugTypeGenerator::convertCharacterType(
         charTy.getLen() * kindMapping.getCharacterBitsize(charTy.getFKind());
   } else {
-    return genPlaceholderType(context);
+    // In some situations, the len of the character is not part of the type but
```
Should we mention "assumed length strings" instead of "some situations" ?

abidh replied:
Thanks for your review. I have updated the comment.

jinisusan reviewed (commented):
Other than the nit mentioned, LGTM

jeanPerier reviewed (commented):
I feel this situation will happen a lot and it would be more generic to add a small fircg.dbg.declare operation that accept any type and that you can generate in DebugTypeGenerator regardless of what is the defining operation (if there is even one).

You could even probably just use llvm.intr.dbg.value already here since FIR uses the same integer types as LLVM, except for the index type. (the llvm.intr.dbg.value accepts any LLVM_Type, which are types that are defined by [this helper](https://github.com/llvm/llvm-project/blob/f25b09199ab8026bf50364cfc8d49a860fb9a760/mlir/lib/Dialect/LLVMIR/IR/LLVMTypes.cpp#L763) and include signeless MLIR integer type and the MLIR floating point types generated by flang).

It would be fine to deal with the index case by generating a convert to i64 I think.

abidh committed:
Handle review comment.

Instead of attaching variable to UnboxCharOp and then generating
DbgValueOp in the Codegen, we now generate it in AddDebugInfo pass. It
removes the need to carry the variable till the Codegen.

abidh committed:
Update comment to address a review comment.

abidh commented:
> I feel this situation will happen a lot and it would be more generic to add a small fircg.dbg.declare operation that accept any type and that you can generate in DebugTypeGenerator regardless of what is the defining operation (if there is even one).
> 
> You could even probably just use llvm.intr.dbg.value already here since FIR uses the same integer types as LLVM, except for the index type. (the llvm.intr.dbg.value accepts any LLVM_Type, which are types that are defined by [this helper](https://github.com/llvm/llvm-project/blob/f25b09199ab8026bf50364cfc8d49a860fb9a760/mlir/lib/Dialect/LLVMIR/IR/LLVMTypes.cpp#L763) and include signeless MLIR integer type and the MLIR floating point types generated by flang).
> 
> It would be fine to deal with the index case by generating a convert to i64 I think.

Thanks for the review. The suggestion to use `llvm.intr.dbg.value` directly here is very useful It will make handling such cases much easier. I have updated the PR accordingly.

jeanPerier commented on flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:
```diff
@@ -289,7 +290,29 @@ mlir::LLVM::DITypeAttr DebugTypeGenerator::convertCharacterType(
+    // variable that will contain that length. This variable is used as
+    // 'stringLength' in DIStringTypeAttr.
+    if (declOp && !declOp.getTypeparams().empty()) {
+      mlir::Operation *op = declOp.getTypeparams()[0].getDefiningOp();
+      if (auto unbox = mlir::dyn_cast_or_null<fir::UnboxCharOp>(op)) {
```
Do you even need to restrict this to unbox?

I think you could just use `declOp.getTypeparams()[0]` directly and this would cover cases where there are no unbox like:

```
subroutine foo(n)
  integer(8) :: n
  character(len=n) :: c
end subroutine
```

jeanPerier reviewed (approved):
Thanks for the update, looks good. You can take my comment to not restrict this to unbox in another path if that seems best to you.
